recipe/build.sh

```diff
@@ -1,43 +1,81 @@
+  export NVCC_PREPEND_FLAGS="-ccbin=$BUILD_PREFIX/bin/${HOST}-clang++ -allow-unsupported-compiler"
 fi
 
-if [ `uname` == Linux ]; then
-    if [ "$target_platform" == "linux-aarch64" ]; then
-        TARGET_ARCH=aarch64
-    elif [ "$target_platform" == "linux-64" ]; then
-        TARGET_ARCH=x86_64
-    fi
-
-    if [ "$PY_VER" == "3.7" ]; then
-        WHL_FILE=https://pypi.org/packages/cp37/c/catboost/catboost-${PKG_VERSION}-cp37-cp37m-manylinux2014_${TARGET_ARCH}.whl
-    elif [ "$PY_VER" == "3.8" ]; then
-        WHL_FILE=https://pypi.org/packages/cp38/c/catboost/catboost-${PKG_VERSION}-cp38-cp38-manylinux2014_${TARGET_ARCH}.whl
-    elif [ "$PY_VER" == "3.9" ]; then
-        WHL_FILE=https://pypi.org/packages/cp39/c/catboost/catboost-${PKG_VERSION}-cp39-cp39-manylinux2014_${TARGET_ARCH}.whl
-    elif [ "$PY_VER" == "3.10" ]; then
-        WHL_FILE=https://pypi.org/packages/cp310/c/catboost/catboost-${PKG_VERSION}-cp310-cp310-manylinux2014_${TARGET_ARCH}.whl
-    elif [ "$PY_VER" == "3.11" ]; then
-        WHL_FILE=https://pypi.org/packages/cp311/c/catboost/catboost-${PKG_VERSION}-cp311-cp311-manylinux2014_${TARGET_ARCH}.whl
-    elif [ "$PY_VER" == "3.12" ]; then
-        WHL_FILE=https://pypi.org/packages/cp312/c/catboost/catboost-${PKG_VERSION}-cp312-cp312-manylinux2014_${TARGET_ARCH}.whl
-    fi
+if [[ "$target_platform" != "$build_platform" ]]; then
+  (
+    mkdir native-build
+    pushd native-build
+
+    mkdir bin
+    ln -sf ${BUILD_PREFIX}/bin/{swig,ragel,yasm} bin/
+
+    export CC=$CC_FOR_BUILD
+    export CXX=$CXX_FOR_BUILD
+    export AR=$($CC_FOR_BUILD -print-prog-name=ar)
+    export NM=$($CC_FOR_BUILD -print-prog-name=nm)
+    export LDFLAGS=${LDFLAGS//$PREFIX/$BUILD_PREFIX}
+    export PKG_CONFIG_PATH=${BUILD_PREFIX}/lib/pkgconfig
+
+    # Unset them as we're ok with builds that are either slow or non-portable
+    unset CFLAGS
+    unset CPPFLAGS
+    unset CXXFLAGS
+    unset CONDA_BUILD_SYSROOT
+
+    cmake \
+     -DCMAKE_PREFIX_PATH=${BUILD_PREFIX} \
+     -DCMAKE_INSTALL_PREFIX=${BUILD_PREFIX} \
+     -DCMAKE_POSITION_INDEPENDENT_CODE=On \
+     -DCMAKE_BUILD_TYPE=Release \
+     -DCATBOOST_COMPONENTS="PYTHON-PACKAGE" \
+    ..
+
+    make -j${CPU_COUNT} _catboost _hnsw
+    popd
+  )
+  export CMAKE_ARGS="${CMAKE_ARGS} -DTOOLS_ROOT=$SRC_DIR/native-build"
 fi
 
-$PYTHON -m pip install --no-deps $WHL_FILE
+if [[ "$cuda_compiler_version" != "None" ]]; then
+  if [[ "$cuda_compiler_version" != "11.8" ]]; then
+    find . -name "CMakeLists*.txt" -type f -print0 | xargs -0 sed -i "s/sm_35/sm_50/g"
+    find . -name "CMakeLists*.txt" -type f -print0 | xargs -0 sed -i "s/compute_35/compute_50/g"
+  fi
+
+  find . -name "CMakeLists*.txt" -type f -print0 | xargs -0 sed -i "s/-lcudart_static/-lcudart/g"
+  find . -name "CMakeLists*.txt" -type f -print0 | xargs -0 sed -i "s/-lcudadevrt/-lcudart/g"
+  find . -name "CMakeLists*.txt" -type f -print0 | xargs -0 sed -i "s/-lculibos/-lcudart/g"
```

Comment:
We discourage linking statically and use shared libraries.

Reply:
1) What do you mean by 'we' in this context? I haven't been able to find a preference for either static of shared dependent libraries linking in the official conda documentation.

I think static linking to CUDA libraries is preferable in this case because users do not need to have CUDA installed in order to use this python package, only nVidia drivers.

2) regarding 'cudadevrt' and 'culibos'.

These are always statically linked.
You can check that they exist only in the static version if you look into the CUDA installation:

```
$ ls -l /usr/local/cuda-11.8/lib64/libcudadevrt*
-rw-r--r-- 1 root root 1021860 сен 21  2022 /usr/local/cuda-11.8/lib64/libcudadevrt.a
$ ls -l /usr/local/cuda-11.8/lib64/libculibos*
-rw-r--r-- 1 root root 30922 сен 21  2022 /usr/local/cuda-11.8/lib64/libculibos.a
$ ls -l /usr/local/cuda-12.0/lib64/libcudadevrt*
-rw-r--r-- 1 root root 1605682 янв 19  2023 /usr/local/cuda-12.0/lib64/libcudadevrt.a
$ ls -l /usr/local/cuda-12.0/lib64/libculibos*
-rw-r--r-- 1 root root 30922 янв 19  2023 /usr/local/cuda-12.0/lib64/libculibos.a
```
And replacing them with `cudart` does not make any sense. These are different libraries.

Reply:
> What do you mean by 'we' in this context? I haven't been able to find a preference for either static of shared dependent libraries linking in the official conda documentation.

From we, I meant conda-forge community. We don't even ship static libraries in most cases. See https://github.com/conda-forge/cfep/blob/main/cfep-18.md for recommendations on shared library.

> I think static linking to CUDA libraries is preferable in this case because users do not need to have CUDA installed in order to use this python package, only nVidia drivers.

We package our own cuda toolkit and therefore users would get a `ibcudart.so` through the conda pacakge manager.

> regarding 'cudadevrt' and 'culibos'. These are always statically linked.

`cudadevrt` and `culibos` are dependencies of `libcudart_static.a` and therefore needs to be linked to if you are using the static version. When using `libcudart.so` it has all the symbols of `libcudart_static.a, libcudadevrt.a, libculibos.a` and therefore all three can be replaced by the shared version.

Reply:
Ok. let's add a comment in the file about it, containing something like that:

```
# Link with shared version of `cudart` library instead of static.
# cudadevrt and culibos are dependencies of libcudart_static.a and therefore needs to be linked to if you are using the static version. When using libcudart.so it has all the symbols of libcudart_static.a, libcudadevrt.a, libculibos.a and therefore all three can be replaced by the shared version.
```